Persistent self-hosted runner: stale tailscale.tgz causes "Destination file path … already exists"

The README says persistent self-hosted runners are supported and that the action "leaves tailscale binaries installed but stops the tailscale background processes" (added in #206). However, on persistent runners where GitHub's `actions/cache` backend doesn't return a cache hit, the action still fails on the second job with:

```
Downloading https://pkgs.tailscale.com/stable/tailscale_<version>_amd64.tgz
Destination file path /home/<user>/.cache/tailscale.tgz already exists
##[error]Destination file path /home/<user>/.cache/tailscale.tgz already exists
```

(repeated 3× by the retry helper)

### Repro

- Persistent self-hosted runner where `$HOME` is preserved across jobs (Aspect Workflows in our case).
- `cache.restoreCache(...)` from `@actions/cache` does not hit (no `Found Tailscale … in cache` line in the logs — it goes straight to `Downloading …`). This happens whenever the runner platform doesn't expose a working GH Actions cache backend.
- Run two consecutive jobs that use this action. The second one fails as above.

Real-world failure: https://github.com/sandbox-quantum/yak/actions/runs/25177419595/job/73813221947 (logs, reproduced on v4.1.2).

### Root cause

[`src/main.ts#L432-L434`](https://github.com/tailscale/github-action/blob/v4.1.2/src/main.ts#L432-L434):

```ts
const tarDest = path.join(xdgCacheDir(), "tailscale.tgz");
fs.mkdirSync(path.dirname(tarDest), { recursive: true });
const tarPath = await tc.downloadTool(downloadUrl, tarDest);
```

`tc.downloadTool` from `@actions/tool-cache` refuses to overwrite an existing destination ([source](https://github.com/actions/toolkit/blob/main/packages/tool-cache/src/tool-cache.ts) — `downloadToolAttempt` throws if `fs.existsSync(dest)`). The action only avoids hitting this code path when the cloud cache restore succeeds, so on runners without a working cache backend the leaked `.tgz` from the previous job collides on every subsequent job.

#206 addressed a related but distinct bug (lingering `tailscaled` daemon → `Text file busy`); the tarball-leak case wasn't covered. The Windows MSI path has a `needsDownload` check ([L538](https://github.com/tailscale/github-action/blob/v4.1.2/src/main.ts#L538)), but the Linux path is unconditional.

### Suggested fix

Either of:

1. Pre-clean the destination before downloading:
   ```ts
   fs.rmSync(tarDest, { force: true });
   const tarPath = await tc.downloadTool(downloadUrl, tarDest);
   ```
2. Pass no `dest` argument so `tc.downloadTool` uses a fresh temp path each time, then delete it after extraction.
3. Mirror the Windows `needsDownload` pattern: if the `.tgz` already exists and its sha256 matches `config.sha256Sum`, skip the download.

Happy to send a PR if option 1 or 3 is acceptable.

### Workaround

Add a step before the action:
```yaml
- run: rm -f "${XDG_CACHE_HOME:-$HOME/.cache}/tailscale.tgz"
```

--- a/src/main.ts
+++ b/src/main.ts
@@ -472,19 +472,46 @@ async function installTailscaleLinux(
 
   // Download and extract
   const downloadUrl = `${baseUrl}/tailscale_${config.resolvedVersion}_${config.arch}.tgz`;
-  logInfo(config.logMode, `Downloading ${downloadUrl}`);
+  const expectedSha = config.sha256Sum.trim().toLowerCase();
 
   const tarDest = path.join(xdgCacheDir(), "tailscale.tgz");
   fs.mkdirSync(path.dirname(tarDest), { recursive: true });
-  const tarPath = await tc.downloadTool(downloadUrl, tarDest);
 
-  // Verify checksum
-  const actualSha = await calculateFileSha256(tarPath);
-  const expectedSha = config.sha256Sum.trim().toLowerCase();
-  logInfo(config.logMode, `Expected sha256: ${expectedSha}`);
-  logInfo(config.logMode, `Actual sha256: ${actualSha}`);
-  if (actualSha !== expectedSha) {
-    throw new Error("SHA256 checksum mismatch");
+  // Check if the tarball already exists with the correct checksum (for
+  // persistent self-hosted runners). tc.downloadTool refuses to overwrite an
+  // existing destination, so a tarball leaked by a previous job would
+  // otherwise fail with "Destination file path already exists" whenever the
+  // GitHub Actions cache backend doesn't return a hit.
+  let tarPath = tarDest;
+  let needsDownload = true;
+  if (fs.existsSync(tarDest)) {
+    const existingSha = await calculateFileSha256(tarDest);
+    if (existingSha === expectedSha) {
+      logInfo(
+        config.logMode,
+        `Using existing tarball at ${tarDest} (checksum verified)`,
+      );
+      needsDownload = false;
+    } else {
+      logInfo(
+        config.logMode,
+        `Existing tarball checksum mismatch, re-downloading`,
+      );
+      fs.unlinkSync(tarDest);
+    }
+  }
+
+  if (needsDownload) {
+    logInfo(config.logMode, `Downloading ${downloadUrl}`);
+    tarPath = await tc.downloadTool(downloadUrl, tarDest);
+
+    // Verify checksum
+    const actualSha = await calculateFileSha256(tarPath);
+    logInfo(config.logMode, `Expected sha256: ${expectedSha}`);
+    logInfo(config.logMode, `Actual sha256: ${actualSha}`);
+    if (actualSha !== expectedSha) {
+      throw new Error("SHA256 checksum mismatch");
+    }
   }
 
   // Extract to tool path

